src/processor/tree/tree_wrapper.rs

```diff
@@ -1,39 +1,41 @@
-use std::{fs, path::Path, str::FromStr};
+use std::{fs, path::Path, str::FromStr, sync::Arc};
 
 use blake2::{Blake2s256, Digest};
 use ethers::types::{Address, H256, U256};
 use eyre::Result;
-use indexmap::IndexSet;
-use state_reconstruct_fetcher::{constants::storage::INITAL_STATE_PATH, types::CommitBlock};
+use state_reconstruct_fetcher::{
+    constants::storage::INITAL_STATE_PATH, database::InnerDB, types::CommitBlock,
+};
+use tokio::sync::Mutex;
 use zksync_merkle_tree::{Database, MerkleTree, RocksDBWrapper};
 
 use super::RootHash;
 
 pub struct TreeWrapper {
     tree: MerkleTree<RocksDBWrapper>,
-    pub index_to_key_map: IndexSet<U256>,
+    snapshot: Arc<Mutex<InnerDB>>,
 }
 
 impl TreeWrapper {
     /// Attempts to create a new [`TreeWrapper`].
-    pub fn new(db_path: &Path, mut index_to_key_map: IndexSet<U256>) -> Result<Self> {
+    pub async fn new(
+        db_path: &Path,
+        snapshot: Arc<Mutex<InnerDB>>,
+        reconstruct: bool,
+    ) -> Result<Self> {
         let db = RocksDBWrapper::new(db_path);
         let mut tree = MerkleTree::new(db);
 
-        // If an existing `index_to_key` mapping was supplied, use that.
-        // Otherwise, reconstruct the genesis state.
-        if index_to_key_map.is_empty() {
-            reconstruct_genesis_state(&mut tree, &mut index_to_key_map, INITAL_STATE_PATH)?;
+        if reconstruct {
+            let mut guard = snapshot.lock().await;
+            reconstruct_genesis_state(&mut tree, &mut *guard, INITAL_STATE_PATH)?;
```

Comment:
Yes, but I think there's little merit in reconstructing that genesis state inside `TreeWrapper::new` anyway. So, what do you think about the following:

1. Make it an associated function.
2. Remove `reconstruct` from parameters.
3. In `TreeProcessor`, if `should_restore`, call `TreeWrapper::reconstruct_genesis_state`

Reply:
I don't like that, because `TreeWrapper` instances aren't useable (`insert_block` must not be called) until `reconstruct_genesis_state` finishes - clearly it's part of the initialization of `TreeWrapper`... Maybe `TreeWrapper` should just own the index to key mapping, and `TreeProcessor` should have the rest of the inner database...

Reply:
BTW if you object specifically to reconstruct, it can probably be reverted to just using the empty state as a flag - I made it explicit mostly because I've been confused... :-)

Reply:
> Maybe TreeWrapper should just own the index to key mapping, and TreeProcessor should have the rest of the inner database...

I think this is the ideal end case, it's just a matter of working out the details. :smiley: 

> BTW if you object specifically to reconstruct [...]

Nah, that's fine by me. For now, I don't have a solution that satisfies both of our conditions, so I guess we can keep as is unless you object?

Reply:
OK - we can test whether this setup can reconstruct the whole chain, and later (if there's time/interest) try to convert to multiple databases (with a common abstraction/base and such)...

Reply:
Sounds good to me! 👍